_posts/2024-03-21-totw-224.md

```diff
@@ -0,0 +1,190 @@
+instance, are as part of a loop from `0` to `size()` and we already know the
+operation will succeed. Therefore, in cases where we already know the bounds
+check will be successful, it's likely that we want the more common
+`operator[]()`.
+
+<pre class="prettyprint code"> {.bad}
+for (int i = 0; i + 1 &lt; vec.size(); ++i) {
+  ProcessPair(vec.at(i), vec.at(i + 1));
+}
+</pre>
+
+becomes
+
+<pre class="prettyprint code"> {.good}
+for (int i = 0; i + 1  &lt; vec.size(); ++i) {
+  ProcessPair(vec[i], vec[i + 1]);
+}
+</pre>
+
+If we do **not** know that the sequence is sufficiently large, is throwing an
+exception the right way to handle that? Usually not. In google3 builds, throwing
+an exception will terminate the program, messily. Many (perhaps most) readers
+won't necessarily spot an innocuously named method like `at()` as a process
+termination risk.
+
+<pre class="prettyprint code"> {.bad}
+std::vector&lt;absl::string_view&gt; tokens = absl::StrSplit(user_string, ByChar(','));
+LOG(INFO) &lt;&lt; "Got leading token " &lt;&lt; tokens.at(0);
+</pre>
+
+is probably better as
+
+<pre class="prettyprint code"> {.good}
+std::vector&lt;absl::string_view&gt; tokens = absl::StrSplit(user_string, ByChar(','));
+if (tokens.empty()) {
+  return absl::InvalidArgumentError("Invalid user_string, expected ','");
+}
+</pre>
+
+or if aborting the program is preferable
+
+<pre class="prettyprint code"> {.good}
+std::vector&lt;absl::string_view&gt; tokens = absl::StrSplit(user_string, ByChar(','));
+CHECK(!tokens.empty()) &lt;&lt; "Invalid user_string "
+                       &lt;&lt; std::quoted(user_string)
+                       &lt;&lt; ", expected at least one ','";
+</pre>
+
+So at least in a google3 context, none of the uses of `at()` are really useful —
+for any given use case, there is a more preferred alternative.
+
+## What About UB?
+
+Unfortunately, reality is hardly so clean as "we know or we don't": we make
+mistakes and code can change over time, invalidating originally correct
+assumptions. Given that humans are fallible, we can *imagine* a use-case for
+`at()`. Specifically, if we are completely consistent in using `at()` instead of
+`operator[]`, we might ensure that even if we're crashing messily (*bad*), we
+don't trigger [undefined behavior (UB)](/tips/labs/ub-and-you) (*worse*).
+
+While we believe "avoid UB" is a very legitimate goal, we still don't endorse
+the use of `at()`, specifically, because of its exception-entangled semantics,
+discussed above. The ideal future solution is a hardened-by-default
+`operator[]`, with compiler optimizations to remove bounds checking, when
+provably safe. The `at()` method is a bad approximation of this solution.
+
+Instead, we encourage users to stick with `operator[]` and reduce exposure to UB
+by other means, including:
+
+*   If your project can afford it, we recommend enabling bounds check in
+    production using
+    [HARDENING_ENABLE_SAFE_LIBCXX](http://go/cc_hardened_binary). Current
```

Comment:
FYI this is an internal golink, could this be replaced with https://github.com/llvm/llvm-project/blob/477b48e8dc6fe2c911332fc40f2fa67f3be016b9/libcxx/docs/Hardening.rst